timespan_types_test.go

```diff
@@ -0,0 +1,198 @@
+package spaniel_test
+
+import (
+	timespan "github.com/senseyeio/spaniel"
+	"testing"
+	"time"
+)
+
+var (
+	t1 = time.Date(2018, 1, 30, 0, 0, 0, 0, time.UTC)
+	t2 = t1.Add(time.Second)
+	t3 = t2.Add(time.Second)
+	t4 = t3.Add(time.Second)
+)
+
+type IntervalTypePair struct {
+	startType timespan.IntervalType
+	endType   timespan.IntervalType
+}
+
+func TestTypedUnion(t *testing.T) {
+
+	for _, tt := range []struct {
+		name       string
+		a, b       timespan.T
+		mergeTypes IntervalTypePair
+	}{
+		{
+			//  (--a--]
+			//        (--b--]
+			name:       "contiguous o/c o/c",
+			a:          timespan.NewEmpty(t1, t2, timespan.Open, timespan.Closed),
+			b:          timespan.NewEmpty(t2, t3, timespan.Open, timespan.Closed),
+			mergeTypes: IntervalTypePair{timespan.Open, timespan.Closed},
+		},
+		{
+			//  [--a--]
+			//        (--b--)
+			name:       "contiguous c/c o/o",
+			a:          timespan.NewEmpty(t1, t2, timespan.Closed, timespan.Closed),
+			b:          timespan.NewEmpty(t2, t3, timespan.Open, timespan.Open),
+			mergeTypes: IntervalTypePair{timespan.Closed, timespan.Open},
+		},
+		{
+			//  [--a--)
+			//        [--b--)
+			name:       "contiguous c/o c/o",
+			a:          timespan.NewEmpty(t1, t2, timespan.Closed, timespan.Open),
+			b:          timespan.NewEmpty(t2, t3, timespan.Closed, timespan.Open),
+			mergeTypes: IntervalTypePair{timespan.Closed, timespan.Open},
+		},
+		{
+			//  [--a--)
+			//    (b]
+			name:       "overlap c/o o/c",
+			a:          timespan.NewEmpty(t1, t4, timespan.Closed, timespan.Open),
+			b:          timespan.NewEmpty(t2, t3, timespan.Open, timespan.Closed),
+			mergeTypes: IntervalTypePair{timespan.Closed, timespan.Open},
+		},
+		{
+			//  [--a--)
+			//    [b)
+			name:       "overlap c/o c/o",
+			a:          timespan.NewEmpty(t1, t4, timespan.Closed, timespan.Open),
+			b:          timespan.NewEmpty(t2, t3, timespan.Closed, timespan.Open),
+			mergeTypes: IntervalTypePair{timespan.Closed, timespan.Open},
+		},
+		{
+			//  (--a--]
+			//  [b]
+			name:       "overlap o/c start instant",
+			a:          timespan.NewEmpty(t1, t4, timespan.Open, timespan.Closed),
+			b:          timespan.NewInstant(t1),
+			mergeTypes: IntervalTypePair{timespan.Closed, timespan.Closed},
+		},
+		{
+			//  [--a--)
+			//       [b]
+			name:       "overlap c/o end instant",
+			a:          timespan.NewEmpty(t1, t4, timespan.Closed, timespan.Open),
+			b:          timespan.NewInstant(t4),
+			mergeTypes: IntervalTypePair{timespan.Closed, timespan.Closed},
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
```

Comment:
You might be able to run all subtest routines in parallel using the `t.Parallel` method since all the cases are self-contained. It could give you a bit of speed up.